filestorage.js

```diff
@@ -0,0 +1,171 @@
+        });
+      } else {
+        cb(null);
+      }
+    });
+  }
+
+  _prepareFilepath(id, cb) {
+    const chunks = common.idToChunks(id);
+    const file = path.join(this.dir, chunks.join('/') + '.' + FS_EXT);
+    this._mkdir(path.dirname(file), err => {
+      cb(err, file);
+    });
+  }
+
+  // Write file to storage
+  //   id - <Object>, common.Uint64, id of file
+  //   data - <string> | <Buffer> | <Uint8Array>, data to be written
+  //   opts - <Object>
+  //     checksum - <string>, checksum type
+  //     dedupHash - <string>, second checksum type
+  //   cb - <Function>, callback
+  //     err - <error>
+  //     checksum - <string>, data checksum
+  //     dedupHash - <string>, second data checksum
+  //     originalSize - <number>, data size
+  write(id, data, opts = {}, cb) {
+    const checksum = this._computeHash(data, opts.checksum || 'CRC32');
+    if (checksum === '') {
+      cb(new Error(`${opts.checksum} is not supported`));
+      return;
+    }
+    const dedupHash = this._computeHash(data, opts.dedupHash || 'SHA256');
+    if (dedupHash === '') {
+      cb(new Error(`${opts.dedupHash} is not supported`));
+      return;
+    }
+    const originalSize = data.length;
+    this._prepareFilepath(id, (err, file) => {
+      if (err) {
+        cb(err);
+      } else {
+        fs.writeFile(file, data, err => {
+          cb(err, checksum, dedupHash, originalSize);
+        });
+      }
+    });
+  }
+
+  // Read file from storage
+  //   id - <Object>, common.Uint64, id of file
+  //   cb - <Function>, callback
+  //     err - <error>
+  //     data - <Buffer>
+  read(id, cb) {
+    const file = path.join(this.dir, common.idToPath(id) + '.' + FS_EXT);
+    fs.readFile(file, cb);
+  }
+
+  // Delete file from storage
+  //   id - <Object>, common.Uint64, id of file
+  //   cb - <Function>, callback
+  //     err - <error>
+  rm(id, cb) {
+    const file = path.join(this.dir, common.idToPath(id) + '.' + FS_EXT);
+    fs.unlink(file, cb);
+  }
+
+  _compress(id, compression, cb) {
+    const file = path.join(this.dir, common.idToPath(id) + '.' + FS_EXT);
+    const filec = file + 'z';
+    switch (compression) {
+      case 'ZIP':
+        zip.compressFile(file, filec)
+          .then(() => {
+            if (fs.statSync(file).size > fs.statSync(filec).size) {
+              fs.rename(filec, file)
+                .then(err => cb(err, true))
+                .catch(err => cb(err, true));
+            } else {
+              fs.unlink(filec, cb);
+            }
+          })
+          .catch(cb);
+        break;
+      case 'GZIP':
+        gzip.compressFile(file, filec)
+          .then(() => {
+            if (fs.statSync(file).size > fs.statSync(filec).size) {
+              fs.rename(filec, file)
+                .then(err => cb(err, true))
+                .catch(err => cb(err, true));
+            } else {
+              fs.unlink(filec, cb);
+            }
+          })
+          .catch(cb);
+        break;
```

Comment:
Could you extract the code into separate functions, it should make the code cleaner.

Also, it looks like both branches are the same and only differ with gzip/zip, could they just be combined to avoid code duplication?